File: src/year2023/day9/day9.rs
//! [Day 9: Mirage Maintenance](https://adventofcode.com/2023/day/9)

struct Puzzle {
    histories: Vec<Vec<i64>>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            histories: data
                .lines()
                .map(|line| {
                    let history: Vec<_> = line
                        .split_whitespace()
                        .map(|x| x.parse::<i64>().unwrap())
                        .collect();
                    history
                })
                .collect(),
        }
    }

    fn history_diffs(history: &[i64]) -> Vec<Vec<i64>> {
        let mut diffs = vec![];

        let mut history = history.to_vec();

        loop {
            // stop when the sequence of differences has only zeros
            if history.iter().min().unwrap() == &0 && history.iter().max().unwrap() == &0 {
                return diffs;
            }

            diffs.push(history.clone());

            history = history
                .iter()
                .zip(history.iter().skip(1))
                .map(|(a, b)| b - a)
                .collect();
        }
    }

    /// Solve part one.
    fn part1(&self) -> i64 {
        let mut result = 0;

        for history in &self.histories {
            let diffs = Self::history_diffs(history);

            // sum of numbers at the right
            let n = diffs.iter().fold(0, |acc, x| acc + x.last().unwrap());

            result += n;
        }

        result
    }

    /// Solve part two.
    fn part2(&self) -> i64 {
        self.histories
            .iter()
            .map(|history| {
                Self::history_diffs(history)
                    .iter()
                    .rev()
                    .fold(0, |acc, x| x.first().unwrap() - acc)
            })
            .sum::<i64>()
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i64, i64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 114);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 2);
    }
}

File: src/year2023/day9/day9.py
#!/usr/bin/env python3
# [Day 9: Mirage Maintenance](https://adventofcode.com/2023/day/9)

import atexit
import sys
import time
from functools import reduce
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


# parse input
lines = data.splitlines()

part1 = 0
part2 = 0

for line in lines:
    line = list(map(int, line.split()))
    diffs = [line]

    while True:
        history = []
        for a, b in zip(line, line[1:]):
            history.append(b - a)

        if set(history) == set([0]):
            break

        diffs.append(history)
        line = history

    right = sum(d[-1] for d in diffs)
    left = reduce(lambda left, d: d[0] - left, reversed(diffs), 0)

    part1 += right
    part2 += left

print(part1)
print(part2)
